File: integration/spark/TestFileLineage.java
import io.openlineage.spark.agent.util.OpenLineageManualEventCreator;
import io.openlineage.client.OpenLineage;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import java.io.File;

public class TestFileLineage {
    public static void main(String[] args) {
        System.out.println("Testing OpenLineageManualEventCreator with file operations...");
        
        SparkSession spark = SparkSession.builder()
            .appName("TestFileLineage")
            .master("local[*]")
            .config("spark.ui.enabled", "false")
            .config("spark.openlineage.namespace", "test_namespace")
            .getOrCreate();

        try {
            // Create temporary directories
            String inputDir = "/tmp/ol_test_input";
            String outputDir = "/tmp/ol_test_output";
            
            // Clean up directories
            deleteDirectory(new File(inputDir));
            deleteDirectory(new File(outputDir));
            
            System.out.println("Creating test input data...");
            
            // Create test data in memory first
            Dataset<Row> testData = spark.range(1, 5)
                .selectExpr("id", "concat('user_', id) as name", "(id * 10) as age");
            
            // Save to parquet (this creates input data)
            testData.write().mode("overwrite").parquet(inputDir);
            System.out.println("✅ Created input data at: " + inputDir);
            
            // Now read from the parquet file (this should be detected as input)
            System.out.println("Reading from input file and transforming...");
            Dataset<Row> inputDF = spark.read().parquet(inputDir);
            
            // Apply transformations
            Dataset<Row> transformedDF = inputDF
                .filter("age > 20")
                .selectExpr("id", "name", "(age + 5) as adjusted_age");
            
            // Save to another parquet file (this should be detected as output)
            transformedDF.write().mode("overwrite").parquet(outputDir);
            System.out.println("✅ Created output data at: " + outputDir);
            
            // Create OpenLineage event from the transformation query execution
            System.out.println("Creating OpenLineage event...");
            OpenLineage.RunEvent event = OpenLineageManualEventCreator.createCompletedEvent(transformedDF.queryExecution());
            
            System.out.println("✅ SUCCESS: Event created!");
            System.out.println("Job Name: " + event.getJob().getName());
            System.out.println("Namespace: " + event.getJob().getNamespace());
            System.out.println("Run ID: " + event.getRun().getRunId());
            
            // Check inputs
            System.out.println("\n📥 Input datasets: " + event.getInputs().size());
            for (OpenLineage.InputDataset input : event.getInputs()) {
                System.out.println("  - " + input.getNamespace() + "/" + input.getName());
                if (input.getFacets() != null && input.getFacets().getSchema() != null) {
                    System.out.println("    Schema fields: " + input.getFacets().getSchema().getFields().size());
                }
            }
            
            // Check outputs
            System.out.println("\n📤 Output datasets: " + event.getOutputs().size());
            for (OpenLineage.OutputDataset output : event.getOutputs()) {
                System.out.println("  - " + output.getNamespace() + "/" + output.getName());
                if (output.getFacets() != null && output.getFacets().getSchema() != null) {
                    System.out.println("    Schema fields: " + output.getFacets().getSchema().getFields().size());
                }
                if (output.getFacets() != null && output.getFacets().getColumnLineage() != null) {
                    System.out.println("    Column lineage present: " + 
                        output.getFacets().getColumnLineage().getFields().getAdditionalProperties().size() + " columns");
                }
            }
            
            // Test in-memory operations (should have no datasets)
            System.out.println("\n🧠 Testing in-memory operations...");
            Dataset<Row> memoryDF = spark.range(1, 5).selectExpr("id", "(id * 2) as doubled");
            OpenLineage.RunEvent memoryEvent = OpenLineageManualEventCreator.createCompletedEvent(memoryDF.queryExecution());
            
            System.out.println("In-memory event inputs: " + memoryEvent.getInputs().size());
            System.out.println("In-memory event outputs: " + memoryEvent.getOutputs().size());
            
            // Clean up
            deleteDirectory(new File(inputDir));
            deleteDirectory(new File(outputDir));
            
        } catch (Exception e) {
            System.err.println("❌ ERROR: " + e.getMessage());
            e.printStackTrace();
        } finally {
            spark.stop();
            System.out.println("\nTest completed.");
        }
    }
    
    private static void deleteDirectory(File dir) {
        if (dir.exists()) {
            File[] files = dir.listFiles();
            if (files != null) {
                for (File file : files) {
                    if (file.isDirectory()) {
                        deleteDirectory(file);
                    } else {
                        file.delete();
                    }
                }
            }
            dir.delete();
        }
    }
} 

File: integration/spark/test_file_lineage.scala
import io.openlineage.spark.agent.util.OpenLineageManualEventCreator
import org.apache.spark.sql.SparkSession
import java.io.File
import java.nio.file.{Files, Paths}

println("Testing OpenLineageManualEventCreator with file-based operations...")

val spark = SparkSession.builder()
  .appName("FileLineageTest")
  .master("local[*]")
  .config("spark.ui.enabled", "false")
  .config("spark.openlineage.namespace", "test_namespace")
  .config("spark.sql.warehouse.dir", "/tmp/spark-warehouse")
  .getOrCreate()

try {
  import spark.implicits._
  
  // Create temporary directories for input and output
  val inputDir = "/tmp/openlineage_test_input"
  val outputDir = "/tmp/openlineage_test_output"
  
  // Clean up any existing directories
  def deleteDirectory(path: String): Unit = {
    val dir = new File(path)
    if (dir.exists()) {
      dir.listFiles().foreach(_.delete())
      dir.delete()
    }
  }
  
  deleteDirectory(inputDir)
  deleteDirectory(outputDir)
  
  println("Creating test data files...")
  
  // Create some test data and save it as Parquet (this will be our input)
  val testData = Seq(
    (1, "Alice", 25, "Engineering"),
    (2, "Bob", 30, "Marketing"), 
    (3, "Charlie", 35, "Engineering"),
    (4, "Diana", 28, "Sales")
  ).toDF("id", "name", "age", "department")
  
  testData.write.mode("overwrite").parquet(inputDir)
  println(s"✅ Created input data at: $inputDir")
  
  // Now read from the parquet file and perform transformations
  println("Reading from input file and performing transformations...")
  val inputDF = spark.read.parquet(inputDir)
  
  val transformedDF = inputDF
    .filter($"department" === "Engineering")
    .select(
      $"id",
      $"name", 
      ($"age" + 5).alias("adjusted_age"),
      $"department".alias("dept")
    )
  
  // Write the result to another parquet file (this will be our output)
  transformedDF.write.mode("overwrite").parquet(outputDir)
  println(s"✅ Created output data at: $outputDir")
  
  // Now create an OpenLineage event from the write operation
  println("Creating OpenLineage event from QueryExecution...")
  val event = OpenLineageManualEventCreator.createCompletedEvent(transformedDF.queryExecution)
  
  println("✅ SUCCESS: File-based lineage event created!")
  println(s"Job Name: ${event.getJob.getName}")
  println(s"Namespace: ${event.getJob.getNamespace}")
  println(s"Event Type: ${event.getEventType}")
  println(s"Run ID: ${event.getRun.getRunId}")
  
  println(s"\n📥 Input datasets: ${event.getInputs.size()}")
  event.getInputs.forEach { input =>
    println(s"  - ${input.getNamespace}/${input.getName}")
    if (input.getFacets.getSchema != null) {
      println(s"    Schema fields: ${input.getFacets.getSchema.getFields.size()}")
      input.getFacets.getSchema.getFields.forEach { field =>
        println(s"      • ${field.getName}: ${field.getType}")
      }
    }
  }
  
  println(s"\n📤 Output datasets: ${event.getOutputs.size()}")
  event.getOutputs.forEach { output =>
    println(s"  - ${output.getNamespace}/${output.getName}")
    if (output.getFacets.getSchema != null) {
      println(s"    Schema fields: ${output.getFacets.getSchema.getFields.size()}")
      output.getFacets.getSchema.getFields.forEach { field =>
        println(s"      • ${field.getName}: ${field.getType}")
      }
    }
    if (output.getFacets.getColumnLineage != null) {
      val columnLineage = output.getFacets.getColumnLineage.getFields.getAdditionalProperties
      println(s"    Column lineage: ${columnLineage.size()} columns")
      columnLineage.forEach { (columnName, lineageInfo) =>
        println(s"      • ${columnName}: ${lineageInfo.getInputFields.size()} input fields")
        lineageInfo.getInputFields.forEach { inputField =>
          println(s"        - ${inputField.getNamespace}/${inputField.getName}/${inputField.getField}")
        }
      }
    }
  }
  
  // Test with an actual write operation to see if we get output datasets
  println("\n🔄 Testing with write operation...")
  val writeOutputDir = "/tmp/openlineage_test_write_output"
  deleteDirectory(writeOutputDir)
  
  val writeOperation = transformedDF.write.mode("overwrite").parquet(writeOutputDir)
  // The write operation itself might generate lineage
  
  // Clean up
  deleteDirectory(inputDir)
  deleteDirectory(outputDir) 
  deleteDirectory(writeOutputDir)
  
} catch {
  case e: Exception =>
    println(s"❌ ERROR: ${e.getMessage}")
    e.printStackTrace()
} finally {
  spark.stop()
  println("\nTest completed.")
} 